File: b3061.py
import sys


sys.setrecursionlimit(10000000)


moves = (0, 1), (0, -1), (1, 0), (-1, 0)


def pos_in_matrix(matrix: list, pos: tuple):
    return len(matrix) > pos[0] >= 0 and len(matrix[0]) > pos[1] >= 0


def apaga_mancha(pele: list, pos: tuple):
    if not pos_in_matrix(pele, pos) or not pele[pos[0]][pos[1]]:
        return False

    pele[pos[0]][pos[1]] = 0

    for move in moves:
        apaga_mancha(pele, (pos[0] + move[0], pos[1] + move[1]))
    return True


def main():
    n, m = tuple(int(number) for number in input().split())
    pele = [[int(item) for item in input().split()] for _ in range(n)]

    manchas = 0

    for i in range(n):
        for j in range(m):
            manchas += apaga_mancha(pele, (i, j))
            
    print(manchas)


if __name__ == '__main__':
    main()

File: b3061.cpp
// https://www.beecrowd.com.br/judge/pt/problems/view/3061

#include <bits/stdc++.h>

using namespace std;

#define _ std::ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
#define dbg(x) cout << #x << " = " << x << endl
#define endl "\n"
#define posInMatrix(matrix, pos) (pos.first < matrix.size() && pos.second < matrix[0].size() && pos.first >= 0 && pos.second >= 0)

typedef unsigned uint;

const vector<pair<short, short>> moves {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

void apagaMancha(vector<vector<short>>& pele, pair<short, short> pos) {
    queue<pair<short, short>> filaApagamento;
    filaApagamento.push(pos);

    pele[pos.first][pos.second] = 0;

    while (!filaApagamento.empty()) {
        pos = filaApagamento.front();
        filaApagamento.pop();

        for (const auto& move : moves) {
            pair<short, short> nextPos {pos.first + move.first, pos.second + move.second};
            if (posInMatrix(pele, nextPos) && pele[nextPos.first][nextPos.second]) {
                filaApagamento.push(nextPos);
                pele[nextPos.first][nextPos.second] = 0;
            }
        }
    }
}

int main() { _
    uint n, m; cin >> n >> m;
    vector<vector<short>> pele(n, vector<short>(m));

    for (int i {0}; i < n; ++i)
        for (int j {0}; j < m; ++j)
            cin >> pele[i][j];

    uint manchas {0};

    for (int i {0}; i < n; ++i)
        for (int j {0}; j < m; ++j)
            if (pele[i][j]) {
                ++manchas;
                apagaMancha(pele, {i, j});
            }

    cout << manchas << endl;

    return 0;
}
